elastalert/test_rule.py

```diff
@@ -261,34 +392,9 @@ def get_id():
             for doc in self.data:
                 doc.update({'_id': doc.get('_id', get_id())})
         else:
-            if args.end:
-                if args.end == 'NOW':
-                    endtime = ts_now()
-                else:
-                    try:
-                        endtime = ts_to_dt(args.end)
-                    except (TypeError, ValueError):
-                        self.handle_error("%s is not a valid ISO8601 timestamp (YYYY-MM-DDTHH:MM:SS+XX:00)" % (args.end))
-                        exit(4)
-            else:
-                endtime = ts_now()
-            if args.start:
-                try:
-                    starttime = ts_to_dt(args.start)
-                except (TypeError, ValueError):
-                    self.handle_error("%s is not a valid ISO8601 timestamp (YYYY-MM-DDTHH:MM:SS+XX:00)" % (args.start))
-                    exit(4)
-            else:
-                # if days given as command line argument
-                if args.days > 0:
-                    starttime = endtime - datetime.timedelta(days=args.days)
-                else:
-                    # if timeframe is given in rule
-                    if 'timeframe' in rule:
-                        starttime = endtime - datetime.timedelta(seconds=rule['timeframe'].total_seconds() * 1.01)
-                    # default is 1 days / 24 hours
-                    else:
-                        starttime = endtime - datetime.timedelta(days=1)
+            # Updating starttime based on timeframe rule
+            if "timeframe" in rule:
```

Comment:
Why do you require timeframe to be define in order to parse the start time? Only one of the several logic branches in parse_starttime() uses timeframe, so my concern is that if someone specifies `days` or `start/end` params but doesn't specify timeframe then it will not yield the expected results.

Reply:
The `timeframe` argument in method `parse_starttime` is not required (line 152):
```
def parse_starttime(self, timeframe=None) -> datetime:
```

The method actually follows the logic previously from this block of code; here more explanations:
1. On first call, at line 178, before rules were applied, so before `timeframe` could be determined, we don’t provide it;
2. And then, here, we can “update” it with `timeframe`, like the previously code was actually doing.

As I understood its logic, do determine `starttime` content, first we read `(self.)args.start`, then `(self.)args.days`, then eventually `timeframe`, and finally using the default of `--days 1` as last resort.

I think I saved the current logic while still giving a refactored version.

What do you think?